### ISSUE TITLE
tile_generator.py encounters IndexError on docker run

### ISSUE THREAD

**Issue by TWCCarlson** [Explv/osrs_map_tiles#2]
Following instructions on https://github.com/Explv/osrs_map_tiles, I ran docker build, which appeared to build correctly. I then pasted the docker run command and encountered the following error:

```
...\CrackTheClue3\osrs_map_tiles-master> docker run -it -v "${pwd}:/repo" map-tile-generator
2022-10-01 00:00:52 INFO Downloading cache & XTEAs
2022-10-01 00:00:56 INFO Cache version: https://archive.openrs2.org/caches/runescape/1117
2022-10-01 00:00:56 INFO Cache upload date: 2022-09-28 10:45:04
2022-10-01 00:00:56 INFO Cache build: ['208']
2022-10-01 00:01:18 INFO Building map base images
Traceback (most recent call last):
  File "./src/tile_generator.py", line 458, in <module>
    main()
  File "./src/tile_generator.py", line 67, in main
    build_full_map_images(cache_dir, xtea_file)
  File "./src/tile_generator.py", line 196, in build_full_map_images
    jar_file = glob.glob("target/*jar-with-dependencies.jar")[0]
IndexError: list index out of range
ERROR conda.cli.main_run:execute(41): `conda run python ./src/tile_generator.py` failed. (See above for error)
```

**Comment by Estebant97** [comment]
+1

**Comment by aaronga97** [comment]
I was able to solve that error by adding the following lines into the `Dockerfile` in line 32:
```
RUN pwd > /tmp/cwd 

RUN cd /runelite && mvn install -DskipTests

RUN cd $(cat /tmp/cwd)
```
Thanks @Bonfire for the hint in https://github.com/Explv/osrs_map_tiles/pull/3#issuecomment-1334538730
I generate the runelite cache doing that (at least the folders). [I followed this guide](https://www.osrsbox.com/blog/2019/06/29/building-and-running-runelite-on-the-terminal-using-maven/) to generate it.

But now I run into the following error:

```
2022-12-08 05:24:29 INFO Downloading cache & XTEAs
2022-12-08 05:24:33 INFO Cache version: https://archive.openrs2.org/caches/runescape/1167
2022-12-08 05:24:33 INFO Cache upload date: 2022-12-07 11:45:06
2022-12-08 05:24:33 INFO Cache build: ['210']
2022-12-08 05:24:51 INFO Building map base images
[main] INFO net.runelite.cache.util.XteaKeyManager - Loaded 1613 keys
[main] INFO net.runelite.cache.MapImageDumper - Map image dimensions: 12544px x 45568px, 4px per map square (1635 MB). Max memory: 8192mb
[main] INFO net.runelite.cache.MapImageDumper - Wrote image /repo/output/generated_images/img-0.png
[main] INFO net.runelite.cache.MapImageDumper - Map image dimensions: 12544px x 45568px, 4px per map square (1635 MB). Max memory: 8192mb
[main] INFO net.runelite.cache.MapImageDumper - Wrote image /repo/output/generated_images/img-1.png
[main] INFO net.runelite.cache.MapImageDumper - Map image dimensions: 12544px x 45568px, 4px per map square (1635 MB). Max memory: 8192mb
[main] INFO net.runelite.cache.MapImageDumper - Wrote image /repo/output/generated_images/img-2.png
[main] INFO net.runelite.cache.MapImageDumper - Map image dimensions: 12544px x 45568px, 4px per map square (1635 MB). Max memory: 8192mb
[main] INFO net.runelite.cache.MapImageDumper - Wrote image /repo/output/generated_images/img-3.png
2022-12-08 05:29:43 INFO Generating tiles
2022-12-08 05:29:43 INFO [Plane: 0]: Generating plane 0
2022-12-08 05:29:43 INFO [Plane: 0]: Loading images into memory
Traceback (most recent call last):
  File "./src/tile_generator.py", line 458, in <module>
    main()
  File "./src/tile_generator.py", line 71, in main
    generate_tiles_for_plane(plane)
  File "./src/tile_generator.py", line 222, in generate_tiles_for_plane
    old_image = pyvips.Image.new_from_file(old_image_location)
  File "/usr/local/bin/miniconda3/envs/vips/lib/python3.7/site-packages/pyvips/vimage.py", line 226, in new_from_file
    raise Error('unable to load from file {0}'.format(vips_filename))
pyvips.error.Error: unable to load from file b'/repo/full_image_0.png'
  VipsForeignLoad: file "/repo/full_image_0.png" does not exist

ERROR conda.cli.main_run:execute(41): `conda run python ./src/tile_generator.py` failed. (See above for error)
```

When I downloaded the repo my Mac literally froze, and every time I `cd` into the directory my mac gets horribly slowed down, so I don't know if that's the reason I'm getting this error or because there's another bug. I'll keep investigating and see if I find out why.

I believe the script might be lookin for the following file https://github.com/Explv/osrs_map_tiles/blob/master/full_image_0.png , but it's looking in path `/repo/full_image_0.png`

I'm not very familiar with docker but I suspect thats  either the last command I added to the Dockerfile `RUN cd $(cat /tmp/cwd)` is not returning to the original location it was at the beginning and somehow that is affecting the `ADD src src` command or it's something else I might be missing.

**Comment by aaronga97** [comment]
I don't believe the generation of the cache to be the problem. I copied Mark's Dockerfile and I still see the exact same error as before. I'll try and find the issue and I'll update here if I find anything.

Does the project work for you @Bonfire?

**Comment by Bonfire** [comment]
The project does not work for me. I was running into the same cache jar issue as what is described here

**Comment by aaronga97** [comment]
## TL;DR
We have to do 2 things, the 1st one solves this issue and the 2nd one solves a bug for MacOS:
1. We have to generate the runelite cache by adding the following line to the `Dockerfile` [file in line 31](https://github.com/Explv/osrs_map_tiles/blob/master/tile_generator/Dockerfile#L31)
    1. `RUN mvn -f /runelite/pom.xml install -DskipTests`. 
 2. And we have to fix the MacOS command, we have to use command substitution `$()` instead of variable expansion `${}`.
    1. `docker run -it -v $(pwd):/repo map-tile-generator` (notice the `$()` instead of `${}`)

## Tracking the issue
Beware, here comes a ~big write up. :octocat:

## Context

The command in the [README.md](https://github.com/Explv/osrs_map_tiles/blob/master/README.md) has instructions on how to run this project. I'm using MacOS so I followed the MacOS section. The last line we have to run is the following:

`docker run -it -v "${pwd}:/repo" map-tile-generator`

This command runs our docker image and it mounts our current working directory (we get it by running `${pwd}`) into the docker image so that we can access our local directory from inside the docker virtualization to save the output files. 

This means that if you are inside the docker image there's going to be a directory called `/repo` that will actually be the directory were you ran the command from. 

I was getting the following error: 

`VipsForeignLoad: file "/repo/full_image_0.png" does not exist`

I tried digging further and the problem was that when I was inside the image and I tried to `ls /repo` it threw me a `file doesn't exist error`, this helped me narrow down the problem to the volume, this mapping from the `docker_image->local_folder` wasn't getting created properly.

After debugging for a while I realized (thanks to the advice from this [stack overflow post](https://stackoverflow.com/questions/55667800/mac-docker-volume-mount-using-osxfs-not-working)) that there's some weird issues going on with `gRPC Fuse` file sharing "protocol?" setting in MacOS, so the first thing I had to do is turn on `osxfs(Legacy)` in `Docker>Settings>General`, following the Stack Overflow advice I verified that the mounting was getting performed correctly and it did.

I ran the command again `docker run -it -v "${pwd}:/repo" map-tile-generator` and it still wasn't working, it didn't make sense. I had already verified that the volume mounting worked, so that wasn't the issue, the only thing left was the command.

I started to test the command and I realized that the mistake was there the entire time in PLAIN SIGHT 😅. If you run the following command in bash `echo "${pwd}"` it prints nothing!

That is because `${}` is for variable expansion in bash and `pwd` is a command not a variable. What we have to use in MacOS and probably in Linux (bash) is command substitution `$()`, and doing that works. Read more about [$() vs ${}](https://unix.stackexchange.com/questions/626752/difference-between-curly-braces-and-brackets-in-command-substitution-and).

## Solution
So going back to the original issue of this link https://github.com/Explv/osrs_map_tiles/issues/2 this issue uncovered the MacOS/Linux command issue.

To solve both issues we have to do the following:

In `tile_generator/Dockerfile` [file in line 31](https://github.com/Explv/osrs_map_tiles/blob/master/tile_generator/Dockerfile#L31) add the following command to generate the cache folders:
`RUN mvn -f /runelite/pom.xml install -DskipTests`

And in MacOS/Linux run the following command instead of the one in the [README.md](https://github.com/Explv/osrs_map_tiles/blob/master/README.md):
`docker run -it -v $(pwd):/repo map-tile-generator`

**Comment by Explv** [comment]
Hey @aaronga97, thank you for taking the time to thoroughly investigate and for your detailed writeup.

I see you didn't create a PR with your fixes, so I will go ahead and push a change myself.

Let me know if there are any further issues, and thanks again for contributing!
